ql/src/java/org/apache/hadoop/hive/ql/io/HiveSequenceFileInputFormat.java

```diff
@@ -44,20 +44,20 @@ public HiveSequenceFileInputFormat() {
     setMinSplitSize(SequenceFile.SYNC_INTERVAL);
   }
 
-  private Set<FileStatus> fileStatuses = null;
+  private ThreadLocal<Set<FileStatus>> fileStatuses = new ThreadLocal<>();
```

Comment:
javadoc :
````
"Each thread holds an implicit reference to its copy of a thread-local variable as long as the thread is alive and the ThreadLocal instance is accessible; after a thread goes away, all of its copies of thread-local instances are subject to garbage collection (unless other references to these copies exist).
````
If your application uses a thread pool, to prevent memory  leaks call the `ThreadLocal.remove()`.

Reply:
Calling `remove()` after the usage of fileStatuses object in the `listStatus()` function. Done.

Reply:
````
public class HiveSequenceFileInputFormat<K extends LongWritable, V extends BytesRefArrayWritable>
  extends SequenceFileInputFormat<K, V> {

  private static final ThreadLocal<Set<FileStatus>> fileStatuses = new ThreadLocal<>();

  public HiveSequenceFileInputFormat() {
    setMinSplitSize(SequenceFile.SYNC_INTERVAL);
  }

  public void setFiles(Set<FileStatus> files) {
    fileStatuses.set(files);
  }

  @Override
  protected FileStatus[] listStatus(JobConf job) throws IOException {
    try {
      if (CollectionUtils.isEmpty(fileStatuses.get())) {
        // In cases where list of files to fetch is not provided we will use SequenceFileInputFormat
        // e.g. SELECT without a job
        return super.listStatus(job);
      }
      return fileStatuses.get().toArray(new FileStatus[0]);
    } finally {
      fileStatuses.remove();
    }
  }
}
````

Reply:
@deniskuzZ I think its called only once. Checked it here - https://github.com/apache/hadoop/blob/trunk/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapred/FileInputFormat.java#L334

Reply:
👍 
1 note, remove this from static var